crates/sui/src/upgrade_compatibility.rs

```diff
@@ -378,36 +381,159 @@ pub(crate) async fn check_compatibility(
         .collect::<Result<Vec<_>, _>>()
         .context("Unable to get existing package")?;
 
-    compare_packages(existing_modules, new_modules)
+    compare_packages(existing_modules, new_package, true)
 }
 
+/// Collect all the errors into a single error message.
 fn compare_packages(
     existing_modules: Vec<CompiledModule>,
-    new_modules: Vec<CompiledModule>,
+    new_package: CompiledPackage,
+    enable_colors: bool,
 ) -> Result<(), Error> {
     // create a map from the new modules
-    let new_modules_map: HashMap<Identifier, CompiledModule> = new_modules
-        .iter()
+    let new_modules_map: HashMap<Identifier, CompiledModule> = new_package
+        .get_modules()
         .map(|m| (m.self_id().name().to_owned(), m.clone()))
         .collect();
 
-    // for each existing find the new one run compatibility check
-    for existing_module in existing_modules {
-        let name = existing_module.self_id().name().to_owned();
-
-        // find the new module with the same name
-        match new_modules_map.get(&name) {
-            Some(new_module) => {
-                Compatibility::upgrade_check().check_with_mode::<CliCompatibilityMode>(
-                    &Module::new(&existing_module),
-                    &Module::new(new_module),
-                )?;
+    let errors: Vec<(Identifier, UpgradeCompatibilityModeError)> = existing_modules
+        .iter()
+        .flat_map(|existing_module| {
+            let name = existing_module.self_id().name().to_owned();
+
+            // find the new module with the same name
+            match new_modules_map.get(&name) {
+                Some(new_module) => {
+                    let compatible = Compatibility::upgrade_check()
+                        .check_with_mode::<CliCompatibilityMode>(
+                            &Module::new(existing_module),
+                            &Module::new(new_module),
+                        );
+                    if let Err(errors) = compatible {
+                        errors.into_iter().map(|e| (name.to_owned(), e)).collect()
+                    } else {
+                        vec![]
+                    }
+                }
+                None => vec![(
+                    name.clone(),
+                    UpgradeCompatibilityModeError::ModuleMissing { name },
+                )],
             }
+        })
+        .collect();
+
+    if errors.is_empty() {
+        return Ok(());
+    }
+
+    let mut files = SimpleFiles::new();
+    let config = codespan_reporting::term::Config::default();
+    let mut writer;
+    if enable_colors {
+        writer = codespan_reporting::term::termcolor::Buffer::ansi();
+    } else {
+        writer = codespan_reporting::term::termcolor::Buffer::no_color();
+    }
```

Comment:
nit: Use the expression form of `if`:

```suggestion
    let writer = if enable_colors {
        Buffer::ansi()
    } else {
        Buffer::no_color()
    }
```